dcos/util.go

```diff
@@ -0,0 +1,297 @@
+	return e.msg
+}
+
+// Util defines an interface to interact with DC/OS cluster via go methods.
+type Util interface {
+	DetectIP() (net.IP, error)
+	Role() (string, error)
+	IsLeader() (bool, error)
+	MesosID() (string, error)
+}
+
+// dcosUtil is implementation of Util interface.
+type dcosUtil struct {
+	client                                             *http.Client
+	detectIPLocation                                   string
+	roleMasterFile, roleAgentFile, roleAgentPublicFile string
+	mesosStateURL                                      string
+}
+
+// NewDCOSUtil returns a new instance of dcosUtil.
+func NewDCOSUtil(client *http.Client, options ...Option) (Util, error) {
+	d := &dcosUtil{
+		client: client,
+	}
+
+	for _, opt := range options {
+		if opt != nil {
+			if err := opt(d); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	// set default `detect_ip` location.
+	if d.detectIPLocation == "" {
+		d.detectIPLocation = FileDetectIP
+	}
+
+	if d.roleMasterFile == "" {
+		d.roleMasterFile = FileMaster
+	}
+
+	if d.roleAgentFile == "" {
+		d.roleAgentFile = FileAgent
+	}
+
+	if d.roleAgentPublicFile == "" {
+		d.roleAgentPublicFile = FileAgentPublic
+	}
+
+	// setup mesos state URL
+	if d.mesosStateURL == "" {
+		stateURL := url.URL{
+			Scheme: "https",
+			Host:   DNSRecordLeader,
+			Path:   "/mesos/state",
+		}
+		d.mesosStateURL = stateURL.String()
+	}
+
+	return d, nil
+}
+
+// DetectIP returns an output from `FileDetectIP` script.
+// This is a nice way of shelling out to `detect_ip` script which handles timeout.
+func (d *dcosUtil) DetectIP() (net.IP, error) {
+	if _, err := os.Stat(d.detectIPLocation); err != nil {
+		return nil, err
+	}
+
+	ce, err := exec.Run("sh", []string{d.detectIPLocation}, exec.Timeout(time.Millisecond*100))
+	if err != nil {
+		return nil, err
+	}
+
+	buf, err := ioutil.ReadAll(ce)
+	if err != nil {
+		return nil, err
+	}
+
+	err = <-ce.Done
+	if err != nil {
+		return nil, err
+	}
+
+	// strip the trailing \n
+	detectedIP := strings.TrimRight(string(buf), "\n")
+	if detectedIP == "" {
+		return nil, ErrDCOSUtil{fmt.Sprintf("command %s return empty output", FileDetectIP)}
+	}
+
+	validIP := net.ParseIP(detectedIP)
+	if validIP == nil {
+		return nil, ErrDCOSUtil{fmt.Sprintf("command %s return invalid IP address %s", FileDetectIP, detectedIP)}
+	}
+
+	return validIP, nil
+}
+
+// Role return a node role.
+func (d *dcosUtil) Role() (string, error) {
+	var isMaster, isAgent, isAgentPublic bool
+	isFile := func(path string, flag *bool) {
+		if _, err := os.Stat(path); err == nil {
+			*flag = true
+		}
+	}
+
+	// iterate over files and get the role.
+	for _, i := range []struct {
+		file string
+		flag *bool
+	}{
+		{
+			file: d.roleMasterFile,
+			flag: &isMaster,
+		},
+		{
+			file: d.roleAgentFile,
+			flag: &isAgent,
+		},
+		{
+			file: d.roleAgentPublicFile,
+			flag: &isAgentPublic,
+		},
+	} {
+		isFile(i.file, i.flag)
+	}
+
+	// make sure we have only one role
```

Comment:
seems clumsy. instead, incorporate the above changes and:
```
if len(roles) > 1 { ...generate error... }
if len(roles) < 1 { ...generate error... }
return roles[0]
```

can write fancier logic once multiple, simultaneous roles are supported